doc/src/manual/environment-variables.md

```diff
@@ -16,8 +16,13 @@ including those which include `JULIA` in their names.
 
 !!! note
 
-    Some variables, such as [`JULIA_NUM_THREADS`](@ref JULIA_NUM_THREADS) and [`JULIA_PROJECT`](@ref JULIA_PROJECT), need to be set before Julia
-    starts, therefore adding these to `~/.julia/config/startup.jl` is too late in the startup process.
+    Some variables, such as [`JULIA_NUM_THREADS`](@ref JULIA_NUM_THREADS) and
+    [`JULIA_PROJECT`](@ref JULIA_PROJECT), need to be set before Julia
+    starts. Adding these to `~/.julia/config/startup.jl` is too late in the startup process.
+
+    Also `__init__()` functions of modules in the sysimage will not see environment variables
+    that are set in `startup.jl` as they are run beforehand.
```

Comment:
```suggestion
    It is recommended to avoid changing environment variables during runtime,
    such as within a `~/.julia/config/startup.jl`.

    One reason is that some julia language variables, such as [`JULIA_NUM_THREADS`](@ref JULIA_NUM_THREADS)
    and [`JULIA_PROJECT`](@ref JULIA_PROJECT), need to be set before Julia starts.

    Similarly, `__init__()` functions of modules in the sysimage are run before `startup.jl`,
    so setting environment variables in a `startup.jl` may be too late for user code.

    Further, changing environment variables during runtime can introduce data races into
    otherwise benign code.
```